File: 2019/haskell/Day22.hs
module Main where

import Data.List (elemIndex, sort, stripPrefix)
import Data.Monoid (Dual(Dual, getDual), Endo(Endo, appEndo))

data Shuffle = Stack | Cut Int | Incr Int deriving Show

parse :: String -> Shuffle
parse "deal into new stack" = Stack
parse s | Just n <- stripPrefix "cut " s = Cut $ read n
        | Just n <- stripPrefix "deal with increment " s = Incr $ read n
parse _ = error "no parse"

stack :: [Integer] -> [Integer]
stack = reverse

cut :: Int -> [Integer] -> [Integer]
cut n cards | n >= 0 = let (h,t) = splitAt n cards
                       in t ++ h
            | otherwise = let (h,t) = splitAt (length cards + n) cards
                          in t ++ h

-- inefficient, but good enough
deal :: Int -> [Integer] -> [Integer]
deal n cards = map snd $ sort $ go 0 cards
  where l = length cards
        go _ [] = []
        go m (c:cs) = (m,c) : go ((m + n) `rem` l) cs

shuffle :: Shuffle -> [Integer] -> [Integer]
shuffle Stack = stack
shuffle (Cut n) = cut n
shuffle (Incr n) = deal n

main :: IO ()
main = do shuffles <- map parse <$> lines <$> readFile "../data/day22"
          let initial = [0..10007-1]
          let final = (getDual $ foldMap (Dual . Endo . shuffle) shuffles) `appEndo ` initial
          let partA = elemIndex 2019 final
          mapM_ print $ partA

File: 2019/rust/src/bin/day22.rs
use std::fs::read_to_string;

enum Shuffle {
    Stack,
    Cut(isize),
    Incr(usize),
}

fn split_prefix<'a, 'b>(pre: &'a str, s: &'b str) -> Option<&'b str> {
    if s.starts_with(pre) {
        return Some(s.split_at(pre.len()).1);
    }
    None
}

fn parse(shuf: &str) -> Shuffle {
    if shuf == "deal into new stack" {
        return Shuffle::Stack;
    }
    match split_prefix("cut ", shuf) {
        Some(n) => return Shuffle::Cut(n.parse().unwrap()),
        None => match split_prefix("deal with increment ", shuf) {
            Some(n) => return Shuffle::Incr(n.parse().unwrap()),
            None => panic!("no parse"),
        },
    }
}

fn do_shuffle(mut deck: Vec<usize>, shuffle: Shuffle) -> Vec<usize> {
    match shuffle {
        Shuffle::Stack => deck.reverse(),
        Shuffle::Cut(n) => {
            let mut t = deck.split_off(if n >= 0 {
                n as usize
            } else {
                deck.len() - (-n) as usize
            });
            t.append(&mut deck);
            deck = t;
        }
        Shuffle::Incr(n) => {
            let p = deck.len();
            let mut new = vec![0; p];
            let mut i = 0;
            for c in deck {
                new[i] = c;
                i += n;
                i %= p;
            }
            deck = new;
        }
    }
    deck
}

fn main() {
    let shufs = read_to_string("../data/day22").unwrap();
    let s = shufs.lines().map(parse);
    let mut deck = (0..10007).collect();
    for x in s {
        deck = do_shuffle(deck, x);
    }
    println!(
        "part a: {}",
        deck.into_iter().position(|c| c == 2019).unwrap()
    );
}
